src/main-cli.cpp

```diff
@@ -85,35 +113,131 @@ int main(int argc, char **argv)
+		while ((ent = readdir (dir)) != NULL) {
+			if (ent->d_name[0]=='.') continue; 
+			std::cout << ent->d_name << std::endl;
+		}
+		closedir (dir);
+	} else {
+		std::cerr << "Could not open presets directory!" << std::endl;
+	}
+}
+
+void process_command(YC20Jack &processor, char *command) {
+	char arg1[128];
+
+	if (strcmp(command, "get_presets\n") == 0) {
+		print_presets();
+	}
+	else if (sscanf(command, "set_preset %[a-zA-Z0-9_:/-.#]\n", arg1) == 1) {
+		std::string dirName(DEFAULT_CONFIG_DIR);
+		std::string fname(arg1);
+		std::string fpath=dirName + "/" + fname;
+		std::cerr << "Loading preset <" << fpath << ">" << std::endl;
+		processor.loadConfiguration(fpath);
+	}
+	else if (strcmp(command, "help\n") == 0) {
+		print_command_help();
+	}
+	else if (strcmp(command, "exit\n") == 0) {
+		exit_cli(0);
+	}
+}
+
+int main(int argc, char **argv)
+{
+	//CLI message: program name and version
+	std::cerr << "Foo-YC20 (CLI) " << get_version() << " (c)Sampo Savolainen 2010" << std::endl << std::endl;
+
+	//Process command line options
+	int force_precalc_osc=0;
+	int exit_after_precalc_osc=0;
+	std::string config_fpath;
+	if (argc > 1) {
+		std::string arg1(argv[1]);
+		if (arg1=="-h") {
+			print_usage();
+			return 0;
+		} else if (arg1=="-l") {
+			print_license();
+			return 0;
+		} else if (arg1=="-f") {
+			force_precalc_osc=1;
+		} else if (arg1=="-g") {
+			force_precalc_osc=1;
+			exit_after_precalc_osc=1;
+		} else {
+			config_fpath=arg1;
+		}
+		if (config_fpath.empty() and argc > 2) {
+			config_fpath=argv[2];
+		}
+	}
 
 	YC20Jack processor;
 	processor.connect();
 
 	dsp *yc20 = createDSP();
-	yc20_precalc_osc *osc = yc20_precalc_oscillators(processor.getSamplerate());
+
+	yc20_precalc_osc *osc=NULL;
+	//Try loading oscillator data from cache file... 
+	std::string fpath=DEFAULT_CONFIG_DIR + "/.precalc_osc.dat";
+	if (!force_precalc_osc) {
+		osc=yc20_load_precalc_osc(processor.getSamplerate(),fpath.c_str());
+	}
+	//If not, precalculate and save to cache ...
+	if (!osc) {
+		osc = yc20_precalc_oscillators(processor.getSamplerate());
+		yc20_save_precalc_osc(osc,fpath.c_str());
+	}
+	//If only cache regeneration option => exit
+	if (exit_after_precalc_osc) {
+		yc20_destroy_oscillators(osc);
+		return 0;
+	}
+	
 	yc20->init(processor.getSamplerate());
 
 	processor.setDSP(yc20);
 	getUserData(yc20)->osc = osc;
 
-	if (argc > 1) {
-		std::string conf(argv[1]);
-		std::cerr << "using configuration file '" << conf << "'" << std::endl;
-		processor.loadConfiguration(conf);
+	// Load configuration
+	if (!config_fpath.empty()) {
+		std::cerr << "Using configuration file '" << config_fpath << "'" << std::endl;
+		processor.loadConfiguration(config_fpath);
 	} else {
 		processor.loadConfiguration();
 	}
+
 	processor.activate();
 
 	run = true;
 	signal(SIGINT, exit_cli);
 
-	while(run) {
-		sleep(100); // ctrl-C interrupts sleep and runs exit_cli() which sets run to false
+	while (run) {
+		char line[128];
+		std::cout << "> ";
```

Comment:
Maybe a `std::flush` will be recommendable here? having some doubt on this.

Reply:
Also, I would prefer to have this interactive command processing as an optional mode. Since there's already command line processing, this could be added as a flag.